mapper/pchain/tx_parser.go

```diff
@@ -0,0 +1,740 @@
+}
+
+func GetAccountsFromUTXOs(hrp string, dependencyTxs map[string]*DependencyTx) (map[string]*types.AccountIdentifier, error) {
+	addresses := make(map[string]*types.AccountIdentifier)
+	for _, dependencyTx := range dependencyTxs {
+		utxoMap := getUTXOMap(dependencyTx)
+
+		for _, utxo := range utxoMap {
+			addressable, ok := utxo.Out.(avax.Addressable)
+			if !ok {
+				return nil, errFailedToGetUTXOAddresses
+			}
+
+			addrs := addressable.Addresses()
+
+			if len(addrs) != 1 {
+				continue
+			}
+
+			addr, err := address.Format(mapper.PChainNetworkIdentifier, hrp, addrs[0])
+			addresses[utxo.UTXOID.String()] = &types.AccountIdentifier{Address: addr}
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return addresses, nil
+}
+
+func GetDependencyTxIDs(tx txs.UnsignedTx) ([]ids.ID, error) {
+	var txIds []ids.ID
+	switch unsignedTx := tx.(type) {
+	case *txs.ExportTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.ImportTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddDelegatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.CreateSubnetTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.CreateChainTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddSubnetValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.RewardValidatorTx:
+		txIds = append(txIds, unsignedTx.TxID)
+	}
+
+	ids.SortIDs(txIds)
+
+	return txIds, nil
+}
+
+func getUniqueTxIds(ins []*avax.TransferableInput) []ids.ID {
+	txnIDs := make(map[string]ids.ID)
+	for _, in := range ins {
+		txnIDs[in.UTXOID.TxID.String()] = in.UTXOID.TxID
+	}
+
+	uniqueTxnIDs := []ids.ID{}
+	for _, txnID := range txnIDs {
+		uniqueTxnIDs = append(uniqueTxnIDs, txnID)
+	}
+	return uniqueTxnIDs
+}
+
+func getUTXOMap(d *DependencyTx) map[uint32]*avax.UTXO {
+	utxos := make(map[uint32]*avax.UTXO)
+
+	if d.Tx != nil {
+		// Generate UTXOs from outputs
+		switch unsignedTx := d.Tx.Unsigned.(type) {
+		case *txs.ExportTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.ImportTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.AddValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.AddDelegatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.CreateSubnetTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.CreateChainTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.AddSubnetValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		}
+	}
+
+	// Add reward UTXOs
+	for _, utxo := range d.RewardUTXOs {
+		utxos[utxo.OutputIndex] = utxo
+	}
+
+	return utxos
+}
+
+func mapUTXOs(txID ids.ID, outs []*avax.TransferableOutput, utxos map[uint32]*avax.UTXO) {
```

Comment:
Why do we take in a map rather than returning one?

Reply:
This function is called twice with different outs (Outs and StakeOuts) for the staking transactions. Passing the same map to both calls instead of dealing with merging the maps in the only caller of this.